src/main/java/de/tum/in/www1/artemis/web/rest/TeamResource.java

```diff
@@ -384,18 +388,28 @@ public TeamResource(TeamRepository teamRepository, TeamService teamService, Team
         // Set teams on exercises
         exercises.forEach(exercise -> exercise.setTeams(Set.of(exerciseTeamMap.get(exercise.getId()))));
 
-        // Fetch participations for team and set the submission count for each participation
-        List<StudentParticipation> participations = participationService.findAllByCourseIdAndTeamShortName(courseId, teamShortName);
+        // Fetch participations for team
+        List<StudentParticipation> participations;
+        if (authCheckService.isAtLeastInstructorInCourse(course, user) || teams.stream().map(Team::getOwner).anyMatch(user::equals)) {
```

Comment:
I don't know if that's possible, but can the teams have different tutors in different exercises?
If so, I believe the `anyMatch` predicate would return too many teams (so maybe `allMatch` should be used?).

Reply:
This is a little tricky. In terms of business logic, all team instances of a team should be assigned to the same tutor for a course. If this were given, it wouldn't matter whether `anyMatch` or `allMatch` is used. However, there is nothing enforcing this in application logic at the moment, so instances of the "same" team (meaning same short name but different team entity in other exercise) could have different tutors if an instructor re-assigns one of the exercise teams to a different tutor.

In this auth check, I decided to opt for `anyMatch` to be a little more lenient if this situation occurs (since the endpoint would otherwise return nothing at all if the tutor is not the owner of one of the instances). Right now, this means: If a user is the team owner for any of the team instances, he gets access to their participations for all exercises.

As long as it's only about reading, I don't think that's a big issue but it actually can be an issue when it comes to the assessment since then more than one tutor would see the assessment button and both would think it's their job to assess this submission.

I think the easiest solution would be if we enforce the rule that the team owner must be the same for all team instances in a course. This is normally given anyway when using the Import Teams functionality but if instructors re-assign tutors afterwards, this change is only performed for the single team instance and not for its clones in other exercises. The solution would be: In the `updateTeam` endpoint, changes to the team owner have to be propagated to all instances of a team.

Reply:
Updates to the team tutor are now propagated to all instances of the same team for a course. I changed the auth check to now use `allMatch` instead so that we definitely don't run into the situation that two tutors are asked to assess the same submission.

When the team update dialog is opened and a different tutor is selected, this warning is now shown:
![OnPaste 20200518-152623](https://user-images.githubusercontent.com/7109225/82221106-d54f4780-991f-11ea-82b9-3610c18ad430.png)

Reply:
Thanks for the clarification!
Your selected solution seems good for me.